spmd/distributed.py

```diff
@@ -0,0 +1,94 @@
+import io
+from dataclasses import replace
+from typing import Any
+
+import torch
+import torch.distributed.distributed_c10d as c10d
+from torch.distributed import ProcessGroup
+from torch.distributed._shard.checkpoint.default_planner import (
+    DefaultLoadPlanner,
+    DefaultSavePlanner,
+)
+from torch.distributed._shard.checkpoint.metadata import (
+    STATE_DICT_TYPE,
+    Metadata,
+    MetadataIndex,
+)
+from torch.distributed._shard.checkpoint.planner import ReadItem, SavePlan
+
+
+def get_ranks(pg: ProcessGroup):
+    """
+    Return an array of global ranks for a given process group.
+    """
+    return [c10d._get_global_rank(pg, i) for i in range(pg.size())]
+
+
+class ProcessGroupAwareSavePlanner(DefaultSavePlanner):
+    """
+    SavePlanner class defining the protocol used by save_state_dict() to plan
+    a process group aware save plan.
+    """
+
+    def create_local_plan(self) -> SavePlan:
+        """
+        Each FSDP process group holds different data but use the same FQN in the state_dict.
+        This will modify the FQN in the save plan to be prefixed by a PG specific string.
+        """
+        save_plan = super().create_local_plan()
+
+        write_item = save_plan.items[0]
+        metadata_index = write_item.index
+        pg_global_ranks = get_ranks(self.state_dict["flat_param"]._process_group)
+        fqn_prefix = "_".join([str(rank) for rank in pg_global_ranks])
+        fqn = fqn_prefix + "_flat_param"
+        metadata_index = replace(metadata_index, fqn=fqn)
+        write_item = replace(write_item, index=metadata_index)
+        save_plan.items[0] = write_item
+
+        return save_plan
+
+    def lookup_object(self, index: MetadataIndex) -> Any:
+        """
+        Remove the PG specific prefix for the fqn to allow object lookup in the state_dict.
+        """
+        if "flat_param" in index.fqn:
```

Comment:
Don't hardcode `flat_param`, record the original names when renaming and use them here. Something like the following:

```
def create_local_plan
  ...
  self.remapping[renamed_key] = original_key

def lookup_object
 ...
 index = replace(index, fqn=self.remapping.get(index.fqn, index.fqn)
```